File: src/YowThi.Erp.Web/src/features/party/farmerMaster.ts
import { getApiJson, postApiCommand } from '../../app/api/apiTransport';
import type { OperationalLocale } from '../../app/i18n/locale';

export { ApiProblemError } from '../../app/api/apiTransport';

export type FarmerMasterStatus = 'all' | 'active' | 'inactive' | 'deleted';

export interface FarmerMasterItem {
  id: string;
  nameZhTw: string | null;
  nameThTh: string | null;
  bankName: string | null;
  bankAccount: string | null;
  phone: string | null;
  address: string | null;
  active: boolean;
  rowVersion: number;
  createdAt: string;
  deletedAt: string | null;
}

export interface FarmerMasterPage {
  items: FarmerMasterItem[];
  nextOffset: number | null;
}

export interface FarmerDraftRequest {
  nameZhTw: string | null;
  nameThTh: string | null;
  bankName: string | null;
  bankAccount: string | null;
  phone: string | null;
  address: string | null;
  active: boolean;
}

export type CreateFarmerRequest = FarmerDraftRequest;

export interface UpdateFarmerRequest extends FarmerDraftRequest {
  expectedRowVersion: number;
}

export interface FarmerWriteResult {
  farmerId: string;
  rowVersion: number;
}

interface FarmerQuery {
  locale: OperationalLocale;
  search?: string;
  status?: FarmerMasterStatus;
  offset?: number;
  limit?: number;
  signal?: AbortSignal;
}

interface FarmerCommandOptions {
  locale: OperationalLocale;
  idempotencyKey: string;
  signal?: AbortSignal;
}

export function listFarmers(query: FarmerQuery): Promise<FarmerMasterPage> {
  const parameters = new URLSearchParams();
  const search = query.search?.trim();
  if (search) parameters.set('search', search);
  if (query.status) parameters.set('status', query.status);
  if (query.offset !== undefined) parameters.set('offset', String(query.offset));
  if (query.limit !== undefined) parameters.set('limit', String(query.limit));
  const suffix = parameters.size === 0 ? '' : `?${parameters.toString()}`;
  return getApiJson<FarmerMasterPage>(`/api/v1/party/farmers${suffix}`, {
    locale: query.locale,
    signal: query.signal,
  });
}

export function createFarmer(
  request: CreateFarmerRequest,
  options: FarmerCommandOptions,
): Promise<FarmerWriteResult> {
  return postApiCommand<CreateFarmerRequest, FarmerWriteResult>(
    '/api/v1/party/farmers',
    request,
    options,
  );
}

export function updateFarmer(
  farmerId: string,
  request: UpdateFarmerRequest,
  options: FarmerCommandOptions,
): Promise<FarmerWriteResult> {
  return postApiCommand<UpdateFarmerRequest, FarmerWriteResult>(
    `/api/v1/party/farmers/${farmerId}/update`,
    request,
    options,
  );
}

File: src/YowThi.Erp.Application/Party/FarmerMaster.cs
using YowThi.Erp.Application.Common.Errors;
using YowThi.Erp.Application.Common.Idempotency;
using YowThi.Erp.Application.Common.Identity;
using YowThi.Erp.Application.Common.Results;

namespace YowThi.Erp.Application.Party;

public sealed record CreateFarmerCommand(
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active);

public sealed record CreateFarmerExecution(
    CommandId CommandId,
    CommandRequestHash RequestHash,
    ActorAccountId ActorAccountId,
    CreateFarmerCommand Command);

public sealed record UpdateFarmerCommand(
    Guid FarmerId,
    long ExpectedRowVersion,
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active);

public sealed record UpdateFarmerExecution(
    CommandId CommandId,
    CommandRequestHash RequestHash,
    ActorAccountId ActorAccountId,
    UpdateFarmerCommand Command);

public sealed record FarmerMasterWriteResult(
    Guid FarmerId,
    long RowVersion);

public interface IFarmerMasterExecutor
{
    ValueTask<ApplicationResult<FarmerMasterWriteResult>> CreateAsync(
        CreateFarmerExecution execution,
        CancellationToken cancellationToken);

    ValueTask<ApplicationResult<FarmerMasterWriteResult>> UpdateAsync(
        UpdateFarmerExecution execution,
        CancellationToken cancellationToken);
}

public enum FarmerMasterStatusFilter
{
    All,
    Active,
    Inactive,
    Deleted,
}

public sealed record FarmerMasterQuery(
    string? Search,
    FarmerMasterStatusFilter Status,
    int Offset,
    int Limit);

public sealed record FarmerMasterItem(
    Guid Id,
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active,
    long RowVersion,
    DateTimeOffset CreatedAt,
    DateTimeOffset? DeletedAt);

public sealed record FarmerMasterPage(
    IReadOnlyList<FarmerMasterItem> Items,
    int? NextOffset);

public interface IFarmerMasterReader
{
    ValueTask<FarmerMasterPage> GetAsync(
        FarmerMasterQuery query,
        CancellationToken cancellationToken);
}

public static class FarmerMasterErrorCodes
{
    public const string InvalidInput = "party.farmer.invalid-input";
    public const string FarmerNotFound = "party.farmer.not-found";
    public const string FarmerDeleted = "party.farmer.deleted";
    public const string StaleRowVersion = "party.farmer.stale-row-version";
    public const string IdempotencyKeyReused = "party.farmer.idempotency-key-reused";
}

public static class FarmerMasterValidation
{
    public static ApplicationError? Validate(CreateFarmerCommand command)
    {
        ArgumentNullException.ThrowIfNull(command);
        return HasVisibleName(command.NameZhTw, command.NameThTh)
            ? null
            : Invalid();
    }

    public static ApplicationError? Validate(UpdateFarmerCommand command)
    {
        ArgumentNullException.ThrowIfNull(command);
        return command.FarmerId != Guid.Empty
            && command.ExpectedRowVersion >= 1
            && HasVisibleName(command.NameZhTw, command.NameThTh)
                ? null
                : Invalid();
    }

    private static bool HasVisibleName(string? nameZhTw, string? nameThTh) =>
        !string.IsNullOrWhiteSpace(nameZhTw) || !string.IsNullOrWhiteSpace(nameThTh);

    private static ApplicationError Invalid() =>
        ApplicationError.Create(
            ApplicationErrorKind.Validation,
            FarmerMasterErrorCodes.InvalidInput);
}
